src/test/script_tapscript_tests.cpp

```diff
@@ -0,0 +1,1711 @@
+                    BOOST_TEST_MESSAGE("control block size " << control_block.size());
+                }
+            }
+
+            // build the witness if necessary
+            if (!m_witness_init) {
+                if (have_control_block) {
+                    // Taproot script path spend
+                    for (const auto& elem : m_initial_witness_stack)
+                        m_witness.stack.push_back(elem);
+                    m_witness.stack.push_back(valtype(m_tapscript.begin(), m_tapscript.end()));
+                    m_witness.stack.push_back(control_block);
+                } else {
+                    // Taproot key path spend
+                    if (!m_witness_signature.empty()) m_witness.stack.push_back(m_witness_signature);
+                }
+                if (!m_annex.empty()) m_witness.stack.push_back(m_annex);
+            }
+
+            if (!m_flags) {
+                m_flags = SCRIPT_VERIFY_SIGPUSHONLY | SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT;
+            }
+
+            SignatureCheckerMock checker_mock(m_checker_validation);
+            CScript script_sig; // must be empty for actual Taproot
+            if (m_p2sh_wrapped) {
+                // But BIP-341 allows all SegWit v1 P2SH-wrapped outputs to pass
+                valtype fake_hash(20, 0x00);
+                script_sig << OP_0 << fake_hash;
+            }
+
+            m_result = VerifyScript(script_sig,
+                                    m_scriptPubKey,
+                                    &m_witness,
+                                    m_flags,
+                                    checker_mock,
+                                    &m_serror);
+
+            m_checker_was_called = checker_mock.CheckerWasCalled();
+
+            return *this;
+        }
+
+        Context& CheckCallSucceeded()
+        {
+            BOOST_CHECK_MESSAGE(m_result,
+                                Descr() << ": VerifyScript succeeded, as expected");
+            BOOST_CHECK_MESSAGE(m_serror == SCRIPT_ERR_OK,
+                                Descr() << ": error code expected OK, actual was "
+                                        << ScriptErrorString(m_serror));
+            return *this;
+        }
+
+        Context& CheckCallFailed(ScriptError expected)
+        {
+            BOOST_CHECK_MESSAGE(!m_result,
+                                Descr() << ": VerifyScript failed, as expected");
+            BOOST_CHECK_MESSAGE(m_serror == expected,
+                                Descr() << ": Error code expected " << ScriptErrorString(expected)
+                                        << ", actual was " << ScriptErrorString(m_serror));
+            return *this;
+        }
+
+        Context& CheckSignatureCheckerWasCalled()
+        {
+            BOOST_CHECK_MESSAGE(m_checker_was_called,
+                                Descr() << ": Schnoor signature checker was called, as expected");
```

Comment:
https://github.com/bitcoin/bitcoin/pull/25097/commits/eeefec343592c6de85048e3b055ba49755cfb072

"Schnoor" misspelling here and elsewhere